### PR TITLE
dependency: Allow formulas to specify :devel as an option for a dependency

### PR DESCRIPTION
**Pull request by dr-impossible**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

**Comment by dr-impossible** [comment]
Currently, it's not possible for a formula to specify the `--devel` option on a dependency. This patch allows that. 

If you define a package with a `devel` block and override the dependencies to also use the `--devel` option, the dependencies fail to build.

Case in point: `gstreamer`. 1.9.1 is the current "devel" version of gstreamer, and all related devel packages in gstreamer require their dependencies to have the same version. For example:

If I add the following to `gst-plugins-base.rb`:
```ruby
  devel do
    url "https://gstreamer.freedesktop.org/src/gst-plugins-base/gst-plugins-base-1.9.1.tar.xz"
    sha256 "8309d1660406274a94bd056f8d805fb35b902208fbb76ff8ca705b37d950fdc9"

    depends_on "gstreamer" => "devel"
  end
```

and the following to `gstreamer.rb`:
```ruby
  devel do
    url "https://gstreamer.freedesktop.org/src/gstreamer/gstreamer-1.9.1.tar.xz"
    sha256 "55304a9e1e8fb5ef82b5b246fef2d9a164eeb0466976a3ad19cfae05b8a94159"
  end
```

and then run the following command:
```bash
brew install gst-plugins-base --devel
```

Homebrew appears to build gstreamer-1.9.1 as a dependency, and is able to run `make` and `make install`, however at some point it selects version 1.8.2 and responds with the error message `Error: empty installation`
Adding the `--debug` option yields the message "Error: You must `brew link gstreamer` before gst-plugins-base can be installed"

**Comment by DomT4** [comment]
Broadly speaking, I think this isn't a good idea, at least until we get `devel` bottles sorted  & shipped. Bottles reduce our maintenance burden significantly, largely because they eliminate the most common breaking factor _(user configuration)_.

A relatively simple change of syntax here somewhat understates how large a change this would be & the impact it would have on maintainers, users & contributors.

There are deeper questions about how to gently degrade if something demands a `:devel` dependency but the formula required doesn't include `devel`. There's also the question of whether it's good UX to allow stable formula to require `:devel` dependencies without user intervention, or whether that's something we'd want to block off or frown upon.

**Comment by dr-impossible** [comment]
I must admit I wasn't aware of the plan to have devel bottles. I'd assumed that using `--devel` came with the similar caveats and pitfalls of using `--head`

I think it's a bit weird to allow a `stable` formula to require a `:devel` dependency. That case would suggest that either the formula isn't really stable, or that the formula for the dependency is outdated.

**Comment by DomT4** [comment]
>  I'd assumed that using --devel came with the similar caveats and pitfalls of using --head

Not as much. We CI test `--devel`, so we know it builds, we know it can be fetched, we know the checksum was correct at time of merge, we know it passes `brew test`, etc. There's no testing on `--HEAD` at all, and likely never will be due to the transient nature of that.

> I think it's a bit weird to allow a stable formula to require a :devel dependency. That case would suggest that either the formula isn't really stable, or that the formula for the dependency is outdated.

It has been requested a few times. The use case is largely where one formula is updated with new releases at a much more rapid cadence than the formula it `depends_on`. I'd lean towards agreeing with you that to me that implies their formula isn't really stable, but that's something that has been requested & we may want to draw a hard line in the sand one way or another if this goes forwards.

**Comment by dr-impossible** [comment]
If you currently override a dependency by adding the `:devel` option within the `devel` block of a formula, `brew audit` will complain but the formula will still build. When you install the formula with `brew install --devel` it appears to download and build the devel version of the dependency, but after the build step, Homebrew checks the prefix path for the stable version of the dependency, which won't be there unless that version was installed as well.

This current behaviour seems a bit broken to me, so I think it would be better for Homebrew to either ignore the `:devel` option on `depends_on` or actually follow it through.

**Comment by dr-impossible** [comment]
For reference, linking to a PR containing an example of the `:devel` option: https://github.com/Homebrew/homebrew-core/pull/3562

**Comment by dr-impossible** [comment]
> There are deeper questions about how to gently degrade if something demands a :devel dependency but the formula required doesn't include devel. 

Based on [this test](https://github.com/Homebrew/brew/blob/028d155e97014076adf4e4dbc7c888fd871de321/Library/Homebrew/test/test_formula_spec_selection.rb#L91), I think if a formula specifies a dependency on `devel` and it's not there, it silently falls back to `stable`:
```ruby
  def test_does_not_raise_for_missing_spec
    f = formula("test", Pathname.new(__FILE__).expand_path, :devel) {
      url "foo-1.0"
      head "foo"
    }

    assert_predicate f, :stable?
  end
```

**Comment by MikeMcQuaid** [comment]
> Based on this test, I think if a formula specifies a dependency on devel and it's not there, it silently falls back to stable:

Yes, it feels like this should 💥 instead.

> I think it's a bit weird to allow a stable formula to require a :devel dependency. That case would suggest that either the formula isn't really stable, or that the formula for the dependency is outdated.

Agreed.

> This current behaviour seems a bit broken to me, so I think it would be better for Homebrew to either ignore the :devel option on depends_on or actually follow it through.

Seems a good enough justification to me.

File: Library/Homebrew/cmd/audit.rb [hunk 1]
```diff
@@ -348,6 +348,7 @@ def audit_deps
         end
 
         dep.options.reject do |opt|
+          next true if opt.name == :devel
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/audit.rb]: Probably want to extend this to if it's `:devel` and the relevant dependency also defines a `devel` spec.

File: Library/Homebrew/dependency.rb [hunk 2]
```diff
@@ -29,7 +29,11 @@ def hash
   end
 
   def to_formula
-    formula = Formulary.factory(name)
+    if tags.include?(:devel)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dependency.rb]: Can make the `formula = if tags`... instead

**Comment by MikeMcQuaid** [comment]
Tests failing on 10.9: http://bot.brew.sh/job/Brew%20Pull%20Requests/1169/version=mavericks/testReport/junit/brew-test-bot/mavericks/tests/

File: Library/Homebrew/cmd/audit.rb (hunk 1 above)
Comment (dr-impossible) [inline Library/Homebrew/cmd/audit.rb]: I'll need to dig a bit deeper to figure out how to do this. Whenever I run `brew audit` on a sample formula that uses this, `spec.build.devel?` is always false unless I run `brew audit --devel`, which isn't what we want. Do you happen to know off-hand if there's somewhere else I should look to detect whether the dependency exists within a `devel` block?

**Comment by ilovezfs** [comment]
How is this going to handle the case where the dependency's development version becomes the stable version (or the stable version becomes newer than the old devel version), the devel spec in the dependency is also upgraded to a new development version, and the dependent formula still needs the original version not the new development version? This idea seems like a can of worms.

**Comment by ilovezfs** [comment]
I believe the answer to my question is that any kind of `=> :devel` thing should have to hard-code a specific version and that version should have to appear in the dependency's devel spec.

Also, I think stable specs should probably never be permitted to depend on `=> :devel` and that should cause both install and audit failures.

File: share/doc/homebrew/Formula-Cookbook.md [hunk 3]
```diff
@@ -136,10 +136,12 @@ A Symbol (e.g. `:x11`) specifies a [`Requirement`](http://www.rubydoc.info/githu
 
 A Hash (e.g. `=>`) specifies a formula dependency with some additional information. Given a single string key, the value can take several forms:
 
-*   a Symbol (currently one of `:build`, `:optional`, `:recommended`).
+*   a Symbol (currently one of `:build`, `:devel`, `:optional`, `:recommended`).
     - `:build` means that dependency is a build-time only dependency so it can
       be skipped when installing from a bottle or when listing missing
       dependencies using `brew missing`.
+    - `:devel` generates an implicit `devel` option, meaning that the development
+      version of the dependency will be installed.
```
Comment (ilovezfs) [inline share/doc/homebrew/Formula-Cookbook.md]: I don't see the use case for this. It smells like feature patching. If the software require's a devel dependency, it should be in an explicit development spec of the main software.

**Comment by ilovezfs** [comment]
The situation this PR seeks to address is pretty rare, and I think vendoring or upstream submodules are usually a better answer for Homebrew. Creating inter-formulae dependencies on `devel` specs would mean `test-bot` would need to understand the scenario and test for breakage in the same way that it tests reverse dependency's for broken `brew test` tests and broken dylibs.

It also complicates the version/upgrade logic as well, beyond just the immediate formula's predicament when the devel spec of the dependency change. For example, does this PR mean I can end up needing multiple versions of the same dependency installed simultaneously? If other software I have installed depends on the stable version, does a devel dependency mean that the software just refuses to install and forces me to choose between mutually exclusive formulae versions, or does it proceed to break silently the other package that needs the stable version of the dependency? What happens at brew upgrade time if a devel version of an installed formula goes from needing a stable version of a dependency to needing a devel version of the dependency, but other packages already installed still need the stable version?

What happens if one of the other dependencies of the formula depending on a devel version depends on a stable version of the same software? Does it have to have its own devel spec added that depends on that devel version? What if its of that software is only optional? Does brew reinstall of that dependency with that option get blocked since it will break the reverse dependency's need for the devel version?

And how will all of this play nicely with the idea in https://github.com/Homebrew/brew/issues/620?

**Comment by ilovezfs** [comment]
>This current behaviour seems a bit broken to me, so I think it would be better for Homebrew to either ignore the :devel option on depends_on or actually follow it through.

This is actually because we basically don't do any option verification. This works without a hitch:
```
diff --git a/Formula/godep.rb b/Formula/godep.rb
index b117cf7..637186a 100644
--- a/Formula/godep.rb
+++ b/Formula/godep.rb
@@ -14,7 +14,7 @@ class Godep < Formula
     sha256 "cdbf4b630a90d8cc4d36076a963916fc979311b2acc413865512b78a41c3aed4" => :mavericks
   end
 
-  depends_on "go"
+  depends_on "go" => [:sparkles, "pokemon", :devel, "newest", "with-lots-of-widgets"]
 
   def install
     ENV["GOPATH"] = buildpath
```
So if that behavior concerns you, option verification seems a bit more relevant that fixating on the string "devel" as being special or justifying any particular change to anything.

Regarding this PR, see https://github.com/Homebrew/homebrew-core/issues/3665 for some issues that arise. For instance, if `godep`'s devel spec depended on `go`'s devel spec, none of the formulae that use `go` and/or `godep` would work on 10.12 since they'd `depend on "go"`, not `depend_on "go" => :devel` and `go`'s stable spec doesn't satisfy the maximum OS requirement.

So would we end up having to add `depends_on "go" => :devel` to all of the formulae that use `go` in a devel spec for each one?

But then how would you build a preexisting devel spec of one of those formulae with the stable version of `go` if you were on 10.11 and didn't want to use new devel `go` but *did* want to use the devel version of that particular formula as you had the day before this change was made?

(No we don't support 10.12 yet, but that's irrelevant; this is just to illustrate what happens when you have a max OS requirement in stable but not in devel.)

File: share/doc/homebrew/Formula-Cookbook.md (hunk 3 above)
Comment (MikeMcQuaid) [inline share/doc/homebrew/Formula-Cookbook.md]: Agreed 👍

**Comment by MikeMcQuaid** [comment]
> I believe the answer to my question is that any kind of => :devel thing should have to hard-code a specific version and that version should have to appear in the dependency's devel spec.

👎 on having specific versions; this wants "version or greater" which we don't really support or want.

There's lots of other questions here that need thought through though.

**Comment by ilovezfs** [comment]
@MikeMcQuaid I think it's relevant here to address the issue raised in https://github.com/Homebrew/homebrew-core/issues/3665#issuecomment-238024925

Is that indeed a bug? Is an up-to-date development version (and also presumably HEAD version) supposed to be able to stand-in for the stable version when fulfilling dependencies and requirements? That seems important to understand in order to be able to assess the meaning of the present PR and how it would impact the ecosystem.

**Comment by MikeMcQuaid** [comment]
@ilovezfs It's not a bug. It used to be `--devel` was applied to everything you were installing at the time but that approach has obvious problems when comparing a dependency that's installed now vs. previously installed.

**Comment by ilovezfs** [comment]
@MikeMcQuaid that's what I suspected. In that case, is it possible to do if MacOS.version == :sierra, stable = Go 1.7rc5 else stable =  "1.6.3"? I think that will fix Homebrew/homebrew-core#3665 without breaking the new approach.

**Comment by MikeMcQuaid** [comment]
@ilovezfs Yeh, could do that. Probably not a bad idea for now at least.

**Comment by ilovezfs** [comment]
I'll simulate it on El Cap using a max OS Yosemite requirement and see if it works.

**Comment by MikeMcQuaid** [comment]
@dr-impossible Any thoughts on addressing these comments (and resolving the conflicts)?

**Comment by ilovezfs** [comment]
@dr-impossible any update here?

**Comment by jeffheifetz** [comment]
I'll comment for @dr-impossible.

> It also complicates the version/upgrade logic as well, beyond just the immediate formula's predicament when the devel spec of the dependency changes. For example, does this PR mean I can end up needing multiple versions of the same dependency installed simultaneously? If other software I have installed depends on the stable version, does a devel dependency mean that the software just refuses to install and forces me to choose between mutually exclusive formulae versions, or does it proceed to break silently the other package that needs the stable version of the dependency? What happens at brew upgrade time if a devel version of an installed formula goes from needing a stable version of a dependency to needing a devel version of the dependency, but other packages already installed still need the stable version?

I believe the project has already taken the approach to never have more than one version of anything linked be it stable, devel or HEAD. But this situation where you have one installed and need another can arise today already and you are forced to choose between mutually exclusive formulae. At least with better dependency listing we can offer better reasoning to the user on which formulae are forcing this situation.

In the example this was drawn from, the installer chooses the wrong version and simply fails.

> What happens if one of the other dependencies of the formula depending on a devel version depends on a stable version of the same software? Does it have to have its own devel spec added that depends on that devel version? What if its use of that software is only optional? Does brew reinstall of that dependency with that option get blocked since it will break the reverse dependency's need for the devel version? Or does the supposedly stable version I'm reinstalling with that option silently start depending on the devel version of the dependency even though stable specs are supposed to only use stable versions of their dependencies?

Again I'd argue that once a user has installed a devel or HEAD formula they can run into issues installing optional formulae. I agree that specifically allowing a devel spec will exacerbate these issues, but it's not as though they don't exist today.


Overall, I think there are a slew of issues that can arise considering the near infinite combinations of packages and dependency trees that can come up. But these issues already can come up today, its simply a question of whether we want to continue to relegate them to the sidelines or not. The problem is exacerbated once we explicitly allow devel dependencies, but hopefully with this better information we can guide users as to how to fix them better.

**Comment by MikeMcQuaid** [comment]
I agree we need to fix this but this PR as-is isn't mergeable, unfortunately and it's been a while since we saw any updates here. Feel free to reopen a new one with conflicts and comments addressed and we can continue discussion there.

**Comment by jeffheifetz** [comment]
I can't seem to find the failing tests anymore, are they secretly cached anywhere?

**Comment by zmwangx** [comment]
No, brew PR builds are kept for 7 days.

**Comment by MikeMcQuaid** [comment]
You should be able to reproduce with `brew tests` locally.

**Pull request closed without merging**